### DISCUSSION TITLE
Use Apicurio SerDe ProtobufKafkaDeserializer against Confluent SerDe Serialized messages

### DISCUSSION THREAD

**Discussion by AndreaScarselliAx** [Apicurio/apicurio-registry#5576] [Q&A]
Hi all,

I am trying to deserialize Protobuf messages with `ProtobufKafkaDeserializer`. These messages have been produced using `io.confluent.kafka.serializers.protobuf.KafkaProtobufSerializer`.

I have already been able to do the same for Avro messages by adding the following to the `AvroKafkaDeserializer` object:

````
srConfigs.putIfAbsent(SerdeConfig.ENABLE_CONFLUENT_ID_HANDLER, Boolean.TRUE.toString());
srConfigs.putIfAbsent(SerdeConfig.USE_ID, IdOption.contentId.name());
````

Doing the same for Protobuf doesn't seem to work, I am actually getting:

`org.apache.kafka.common.config.ConfigException: Invalid value java.lang.reflect.InvocationTargetException for configuration Not a valid protobuf builder`

Is there anything else needed for Protobuf to work?

Thank you and kind regards,
Andrea

**Comment by Ganesh-403** [comment]
The Avro case works because Apicurio provides explicit compatibility support for Confluent's wire format through:

```java
SerdeConfig.ENABLE_CONFLUENT_ID_HANDLER
```

(or the equivalent `apicurio.registry.as-confluent=true` configuration). This allows the deserializer to interpret the Confluent-encoded schema ID correctly. :contentReference[oaicite:0]{index=0}

However, **Protobuf has an additional requirement beyond reading the schema ID**.

The Apicurio documentation notes that `ProtobufKafkaDeserializer` deserializes messages into a **`DynamicMessage`**, not into your generated Protobuf message class. It also requires the corresponding Protobuf schema to reconstruct the message. :contentReference[oaicite:1]{index=1}

The exception:

```text
ConfigException: Invalid value java.lang.reflect.InvocationTargetException
Not a valid protobuf builder
```

suggests that the deserializer is trying to instantiate or use a Protobuf message builder that isn't compatible with the incoming message.

From what I can tell, I couldn't find any documentation stating that `ENABLE_CONFLUENT_ID_HANDLER` alone enables interoperability with **Confluent's `KafkaProtobufSerializer`** in the same way it does for Avro. The documentation only describes Confluent-compatible ID handling—it doesn't mention full compatibility with Confluent's Protobuf serializer implementation. :contentReference[oaicite:2]{index=2}

A few things I'd check:

1. **How are you configuring `ProtobufKafkaDeserializer`?**
   - Are you deserializing into a `DynamicMessage`, or are you trying to obtain a generated Protobuf class?
2. **Which versions** of:
   - Apicurio Registry client/SerDe
   - Confluent serializers
   - `protobuf-java`
   are you using?
3. Could you share the complete stack trace? In particular, the frames immediately following `Not a valid protobuf builder` would help identify which builder class is being instantiated.

If your goal is to consume messages produced by Confluent's `KafkaProtobufSerializer`, this may be a limitation (or a bug) in the current interoperability support rather than a missing configuration option. I couldn't find documentation confirming that this cross-serializer scenario is officially supported for Protobuf.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
@AndreaScarselliAx
